### PR TITLE
Vision function calling training

### PR DESCRIPTION
**Pull request by khai-meetkai**
+ Add prompt_template, training, and server inference for vision function calling

### CODE REVIEW

File: functionary/train_vision/README.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# Install Dependencies
+```shell
+pip install accelerate==0.27.2 bitsandbytes==0.41.1 scipy==1.11.3 sentencepiece==0.1.99 packaging==23.1 ninja==1.11.1 einops==0.7.0 wandb==0.15.11 jsonref==1.1.0 deepspeed==0.14.2 typer==0.9.0 tensorboard==2.15.1 wheel==0.42.0 aenum==3.1.15 transformers==4.42.3 flash-attn git+https://github.com/LLaVA-VL/LLaVA-NeXT.git
```
[the remark this answers is not in the document]
Comment (musab-mk) [reply functionary/train_vision/README.md:3]: We can convert this into requirements.txt file too

File: functionary/train_vision/llava_dataset.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
+from torch.utils.data import Dataset, Sampler
+import transformers
+from typing import Dict, Any, Optional
+import torch
+from PIL import Image
+import random
+random.seed(100)
```
[the remark this answers is not in the document]
Comment (musab-mk) [reply functionary/train_vision/llava_dataset.py:7]: And document it

File: functionary/train_vision/train.py [hunk 3]
```diff
@@ -0,0 +1,539 @@
...
+    if num_new_tokens > 0:
+        input_embeddings = model.get_input_embeddings().weight.data
+        output_embeddings = model.get_output_embeddings().weight.data
+
+        input_embeddings_avg = input_embeddings[:-num_new_tokens].mean(
+            dim=0, keepdim=True
+        )
+        output_embeddings_avg = output_embeddings[:-num_new_tokens].mean(
+            dim=0, keepdim=True
+        )
```
Comment (musab-mk) [inline functionary/train_vision/train.py:197]: Why resize is needed for this model? Could you add an explanation

File: functionary/train_vision/train.py [hunk 4]
```diff
@@ -0,0 +1,539 @@
...
+        training_args.remove_unused_columns,
+    )
+
+    compute_dtype = (
+        torch.float16
+        if training_args.fp16
+        else (torch.bfloat16 if training_args.bf16 else torch.float32)
+    )
+
+    # Copy this from: https://github.com/LLaVA-VL/LLaVA-NeXT/blob/inference/llava/model/builder.py#L169
```
Comment (musab-mk) [inline functionary/train_vision/train.py:248]: Could you use permanent URL instead of latest main branch's Line number

File: functionary/train_vision/train.py [hunk 5]
```diff
@@ -0,0 +1,539 @@
...
+        eval_dataset = LazyVisionDataset(raw_eval_ds, tokenizer, data_args.pad_img_path)
+
+        if torch.distributed.get_rank() == 0:
+            print(f"Eval Data Loaded: #{len(eval_dataset)}")
+
+    print_rank0("***** HERE ARE SOME EXAMPLES FROM TRAINING ****")
+    # print_some_examples(train_dataset, tokenizer)
+
+    print_rank0("***** HERE ARE SOME EXAMPLES FROM EVALUATION ***")
+    # print_some_examples(eval_dataset, tokenizer)
```
Comment (musab-mk) [inline functionary/train_vision/train.py:326]: We can remove this part if we dont print

File: functionary/train_vision/train.py [hunk 6]
```diff
@@ -0,0 +1,539 @@
...
+            "accuracy_first_token": first_token_correct_count / first_token_total_count,
+            "total_number_first_token": first_token_total_count,
+        }
+
+        for token_id, stat in sorted(
+            first_token_label_dic.items(), key=lambda x: -x[1]["total"]
+        )[:5]:
+            token = tokenizer.decode([token_id])
+            metrics[f"accuracy_first_token_{token}"] = stat["correct"] / stat["total"]
+            metrics[f"accuracy_first_token_{token}_total"] = stat["total"]
```
Comment (musab-mk) [inline functionary/train_vision/train.py:400]: Let's use "accuracy recipient token".  to make it more clear.

Comment (musab-mk) [reply functionary/train_vision/train.py:400]: Since "first token" might be ambiguous.

Comment (musab-mk) [reply functionary/train_vision/train.py:400]: Also, adding "all" token accuracy is helpful. For differentiating between function call's function name accuracy and followup accuracy. Let's add that accuracy too. You may need to manually extract token id of "all"

File: functionary/train_vision/train.py [hunk 7]
```diff
@@ -0,0 +1,539 @@
...
+        else:
+            image_tensor = None
+        # if type(image_tensor) is not list:
+        #    image_tensor = image_tensor.to(model.dtype)
+
+        result["images"] = image_tensor
+        result["image_sizes"] = image_sizes
+        return result
+
+    class CustomizedTrainer(Trainer):
```
Comment (musab-mk) [inline functionary/train_vision/train.py:460]: Could you add more descriptive name than "Customized". Maybe FunctionAccuracyTrackingTrainer or something like that. And let's add one or two sentence of explanation why we customized the trainer.

**Review by musab-mk — CHANGES_REQUESTED** [review]

File: functionary/inference_vision.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    prepare_messages_for_inference,
+    get_images_from_messages,
+)
+from typing import Any, List, Optional
+from llava.mm_utils import process_images
+import torch
+from functionary.inference_utils import analyze_tools_and_tool_choice
+
+
+def generate_message(
```
[the remark this answers is not in the document]
Comment (khai-meetkai) [reply functionary/inference_vision.py:19]: Ok, sure !

File: functionary/prompt_template/llava_prompt_template.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        #     content = f"name={tool_name}\n{content}"
+
+        prompt_template = (
+            f"{self.start_header}{role}{self.end_header}\n\n"
+            + "{text}"
+            + self.eos_token
+        )
+
+        if role == "user":  # Check if contain uploaded image or not
+            if type(content) is list:
```
[the remark this answers is not in the document]
Comment (khai-meetkai) [reply functionary/prompt_template/llava_prompt_template.py:29]: Ok, sure !

File: functionary/prompt_template/llava_prompt_template.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+                content = prompt_utils.get_content_str_from_multi_modal_input(
+                    message["content"], self.image_token
+                )
+
+        if role in ["user", "system", "tool"]:
```
[the remark this answers is not in the document]
Comment (khai-meetkai) [reply functionary/prompt_template/llava_prompt_template.py:34]: Oh I think openAI client already made sure that role in user, system, tool, assistant.  By the way, I will raise an exception here

File: functionary/prompt_template/llava_prompt_template.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+            return prompt_template.format(text=content)
+
+        assert role == "assistant", f"role must be assistant, but: {role}"
+        tool_calls = message.get("tool_calls", [])
+        if tool_calls is None:
+            tool_calls = []
```
[the remark this answers is not in the document]
Comment (khai-meetkai) [reply functionary/prompt_template/llava_prompt_template.py:40]: Oh, I don't know this, thank you

File: functionary/prompt_template/llava_prompt_template.py [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+        for tool_call in tool_calls:
+            arguments = tool_call["function"]["arguments"]
+            tool_name = tool_call["function"]["name"]
+            tool_prompt = f"{tool_name}\n{arguments}"
+            tool_call_prompts.append(tool_prompt)
+
+        total_content = self.function_separator + self.function_separator.join(
+            tool_call_prompts
+        )
+        return prompt_template.format(text=total_content)
```
[the remark this answers is not in the document]
Comment (khai-meetkai) [reply functionary/prompt_template/llava_prompt_template.py:60]: Actually this is the copy from v3.llama, I think I will re-write this to inherit the implementation from Llama3TemplateV3

File: functionary/prompt_template/prompt_utils.py [hunk 13]
```diff
@@ -203,3 +206,55 @@ def reorder_tool_messages_by_tool_call_ids(messages: List[Dict]) -> List[Dict]:
...
+def get_content_str_from_multi_modal_input(
+    content: List[Dict], image_token: str
+) -> str:
+    result = ""
+    for item in content:
+        if item["type"] == "text":
+            result += item["text"]
+        elif item["type"] == "image_url":
+            result += image_token
+    return result
```
[the remark this answers is not in the document]
Comment (khai-meetkai) [reply functionary/prompt_template/prompt_utils.py:220]: Sure, I will rename to this: stringify_content_with_images

File: functionary/prompt_template/prompt_utils.py [hunk 14]
```diff
@@ -203,3 +206,55 @@ def reorder_tool_messages_by_tool_call_ids(messages: List[Dict]) -> List[Dict]:
...
+def get_images_from_messages(messages: List[Dict]) -> List:
+    result = []
+    for message in messages:
+        if message["role"] == "user":
+            if type(message["content"]) is list:
+                result.extend(get_images_from_content(message["content"]))
+    return result
+
+
+def get_images_from_content(content: List[Dict]) -> List:
```
[the remark this answers is not in the document]
Comment (khai-meetkai) [reply functionary/prompt_template/prompt_utils.py:232]: Ok, sure !

File: functionary/train/README.md [hunk 15]
```diff
@@ -7,7 +7,7 @@ python3 -m venv venv && source venv/bin/activate
 pip3 install torch==2.0.1 torchvision==0.15.2 torchaudio==2.0.2 --index-url https://download.pytorch.org/whl/cu118
 
 # Install Dependencies
-pip install accelerate==0.27.2 bitsandbytes==0.41.1 scipy==1.11.3 sentencepiece==0.1.99 packaging==23.1 ninja==1.11.1 einops==0.7.0 wandb==0.15.11 jsonref==1.1.0 deepspeed==0.14.2 typer==0.9.0 tensorboard==2.15.1 wheel==0.42.0 aenum==3.1.15 git+https://github.com/huggingface/transformers.git flash-attn==2.5.9
+pip install accelerate==0.27.2 bitsandbytes==0.41.1 scipy==1.11.3 sentencepiece==0.1.99 packaging==23.1 ninja==1.11.1 einops==0.7.0 wandb==0.15.11 jsonref==1.1.0 deepspeed==0.14.2 typer==0.9.0 tensorboard==2.15.1 wheel==0.42.0 aenum==3.1.15 git+https://github.com/huggingface/transformers.git flash-attn
```
[the remark this answers is not in the document]
Comment (khai-meetkai) [reply functionary/train/README.md:10]: Sure !

File: functionary/train_vision/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (khai-meetkai) [reply functionary/train_vision/README.md:3]: Sure

File: functionary/train_vision/README.md [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+    --gradient_accumulation_steps 1 \
+    --eval_accumulation_steps 1 \
+    --evaluation_strategy "steps" \
+    --eval_steps 100 \
+    --save_strategy "no" \
+    --save_steps 100 \
+    --save_total_limit 3 \
+    --logging_steps 1 \
+    --learning_rate 1e-5 \
+    --weight_decay 0. \
```
[the remark this answers is not in the document]
Comment (khai-meetkai) [reply functionary/train_vision/README.md:27]: Ok, Sure !

File: functionary/train_vision/README.md [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+    --warmup_ratio 0.03 \
+    --lr_scheduler_type "cosine" \
+    --tf32 True \
+    --model_max_length 4096 \
```
[the remark this answers is not in the document]
Comment (khai-meetkai) [reply functionary/train_vision/README.md:31]: Yes, when I ran the real training, I used 8k, 4k was only set when I cleaned the code and wanted to run the training to make sure it still worked after cleaning

File: functionary/train_vision/assert_injecting_virtual_img_token.py
[the remark this answers is not in the document]
Comment (khai-meetkai) [reply functionary/train_vision/assert_injecting_virtual_img_token.py:1]: Running this assert file is **optional** to make sure that **adding virtual_img_token won't change the loss if someone is skeptical of it**. This is also to make sure that the way I implemented adding virtual img token is correct. 
We cannot put this in the **tests** because we **don't need to run it again and again**, it is fixed, and it **needs GPU** to run the test.

File: functionary/train_vision/llava_dataset.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (khai-meetkai) [reply functionary/train_vision/llava_dataset.py:7]: Sure !

File: functionary/train_vision/models/modeling_llava.py [hunk 18]
```diff
@@ -0,0 +1,394 @@
...
+    DEFAULT_IM_START_TOKEN,
+    DEFAULT_IM_END_TOKEN,
+)
+from llava.mm_utils import get_anyres_image_grid_shape
+from llava.model.llava_arch import unpad_image
+from typing import Optional, List, Union, Tuple
+from transformers.modeling_outputs import CausalLMOutputWithPast
+
+
+class FixedLlavaLlamaForCausalLM(LlavaLlamaForCausalLM):
```
[the remark this answers is not in the document]
Comment (khai-meetkai) [reply functionary/train_vision/models/modeling_llava.py:17]: Yes, but it might be confused as the subclass name is identical to the parent class name

File: functionary/train_vision/models/modeling_llava.py [hunk 19]
```diff
@@ -0,0 +1,394 @@
...
+
```
[the remark this answers is not in the document]
Comment (khai-meetkai) [reply functionary/train_vision/models/modeling_llava.py:18]: Sure !

File: functionary/train_vision/train.py [hunk 20]
```diff
@@ -0,0 +1,539 @@
...
+        },
+    )
+
+
+@dataclass
+class TrainingArguments(transformers.TrainingArguments):
+    cache_dir: Optional[str] = field(default=None)
+    optim: str = field(default="adamw_8bit")
+    model_max_length: int = field(
+        default=4096,
```
[the remark this answers is not in the document]
Comment (khai-meetkai) [reply functionary/train_vision/train.py:123]: Sure

File: functionary/train_vision/train.py (hunk 3 above)
Comment (khai-meetkai) [reply functionary/train_vision/train.py:197]: Oh this is **just in case we add new tokens to the tokenizer** In the future we may add new tokens, currently we don't. If we don't add new tokens, this chunk of code has no effect

File: functionary/train_vision/train.py (hunk 4 above)
Comment (khai-meetkai) [reply functionary/train_vision/train.py:248]: what do you mean by permanent URL?  You mean the URL with the commit adding this code?

File: functionary/train_vision/train.py (hunk 5 above)
Comment (khai-meetkai) [reply functionary/train_vision/train.py:326]: Sure

File: functionary/train_vision/train.py (hunk 6 above)
Comment (khai-meetkai) [reply functionary/train_vision/train.py:400]: Here I added top-5 top 5 function_names that **appear the most in the data**, so "all" is always expected to be in the result: `accuracy_recipient_token_all`. But I will fix this to include: "all" even it is out of top-5

File: functionary/train_vision/train.py (hunk 7 above)
Comment (khai-meetkai) [reply functionary/train_vision/train.py:460]: Sure

File: functionary/train_vision/models/modeling_llava.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (musab-mk) [reply functionary/train_vision/models/modeling_llava.py:18]: Okay I didnt realize this.

**Review by musab-mk — APPROVED** [review]
